File: packages/sdk-go/usage.go
package driftstack

import (
	"context"
	"net/url"
	"strconv"
)

// UsageResource handles /v1/usage + /v1/usage/series.
type UsageResource struct {
	client *Client
}

// CurrentPeriod returns the calendar-month-UTC totals + tier quotas
// for the current account.
func (r *UsageResource) CurrentPeriod(ctx context.Context) (*UsagePeriodSummary, error) {
	var out UsagePeriodSummary
	if err := r.client.do(ctx, requestOptions{
		method: "GET",
		path:   "/v1/usage",
		out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

// UsageDailyBucket — V-452 single-day bucket on the time series.
// `Totals` keys are the UsageRecordType enum values (e.g.
// "session_minute", "navigate", "interact", etc.).
type UsageDailyBucket struct {
	Date   string         `json:"date"` // YYYY-MM-DD
	Totals map[string]int `json:"totals"`
}

// UsageSeriesResponse — V-452 daily-bucketed time series.
type UsageSeriesResponse struct {
	FromDate string             `json:"from_date"`
	ToDate   string             `json:"to_date"`
	Buckets  []UsageDailyBucket `json:"buckets"`
}

// Series — V-452 daily-bucketed usage time series. `days` is 1-90;
// default 30 server-side when 0 / negative.
func (r *UsageResource) Series(ctx context.Context, days int) (*UsageSeriesResponse, error) {
	var out UsageSeriesResponse
	q := url.Values{}
	if days > 0 {
		q.Set("days", strconv.Itoa(days))
	}
	if err := r.client.do(ctx, requestOptions{
		method: "GET",
		path:   "/v1/usage/series",
		query:  q,
		out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

File: packages/sdk-python/src/driftstack/resources/usage.py
"""Usage resource — /v1/usage + /v1/usage/series."""

from __future__ import annotations

from urllib.parse import urlencode

from pydantic import BaseModel

from driftstack._generated.models import UsagePeriodSummary
from driftstack.http import AsyncHttpClient, HttpClient, parse_model


class UsageDailyBucket(BaseModel):
    """One UTC-day bucket in a :class:`UsageSeriesResponse`.

    Mirrors api-types ``UsageDailyBucketSchema``. ``date`` is
    ``YYYY-MM-DD``; ``totals`` maps each usage-record type to its count
    for the day (zero-usage days are present with an empty ``totals``).
    """

    date: str
    totals: dict[str, int]


class UsageSeriesResponse(BaseModel):
    """Response shape for ``GET /v1/usage/series`` (V-452 / V-170).

    Mirrors api-types ``UsageSeriesResponseSchema`` — TS/Go already
    return this typed. ``to_date`` is exclusive, ``from_date``
    inclusive; ``buckets`` is contiguous, one per UTC day.
    """

    from_date: str
    to_date: str
    buckets: list[UsageDailyBucket]


class UsageResource:
    """Synchronous usage resource."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def current_period(self) -> UsagePeriodSummary:
        """Current calendar-month UTC totals + tier quotas."""
        data = self._http.request("GET", "/v1/usage")
        return parse_model(UsagePeriodSummary, data)

    def series(self, *, days: int | None = None) -> UsageSeriesResponse:
        """V-452 — daily-bucketed usage time series. ``days`` is 1-90;
        default 30. Returns ``{from_date, to_date, buckets}``.
        """
        path = "/v1/usage/series"
        if days is not None:
            path = f"{path}?{urlencode({'days': days})}"
        data = self._http.request("GET", path)
        return parse_model(UsageSeriesResponse, data)


class AsyncUsageResource:
    """Async usage resource."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def current_period(self) -> UsagePeriodSummary:
        data = await self._http.request("GET", "/v1/usage")
        return parse_model(UsagePeriodSummary, data)

    async def series(self, *, days: int | None = None) -> UsageSeriesResponse:
        path = "/v1/usage/series"
        if days is not None:
            path = f"{path}?{urlencode({'days': days})}"
        data = await self._http.request("GET", path)
        return parse_model(UsageSeriesResponse, data)
